File: src/containers/students.ts
import type { Gift, Skill, Student } from "@prisma/client";
import { similarity } from "../utils/similarity";
import { db } from "../db/client";

export type DetailedStudent = Student & {
  baseVariant?: Student | null;
  skills: Skill[];
  giftsAdored: Gift[];
  giftsLoved: Gift[];
  giftsLiked: Gift[];
};

export class StudentContainer {
  private students: Map<string, DetailedStudent> = new Map();

  async bootstrap(): Promise<void> {
    await this.reload();
  }

  async reload(): Promise<void> {
    this.students.clear();

    const students = await db.student.findMany({
      include: {
        baseVariant: true,
        skills: true,
        giftsAdored: true,
        giftsLoved: true,
        giftsLiked: true,
      },
    });

    for (const student of students) {
      this.addStudent(student.id, student);
    }
  }

  addStudent(key: string, student: DetailedStudent): void {
    this.students.set(key, student);
  }

  getStudent(key: string): DetailedStudent | null {
    if (this.students.has(key)) {
      return this.students.get(key) ?? null;
    }

    return null;
  }

  getStudents(): DetailedStudent[] {
    return Array.from(this.students.values());
  }

  getStudentsWhere(
    predicate: (student: DetailedStudent) => boolean,
  ): DetailedStudent[] {
    return this.getStudents().filter(predicate);
  }

  getStudentKeys(): string[] {
    return Array.from(this.students.keys());
  }

  getStudentKeysWhere(
    predicate: (student: DetailedStudent) => boolean,
  ): string[] {
    return Array.from(this.students.entries())
      .filter(([_, student]) => predicate(student))
      .map(([key, _]) => key);
  }

  all(): Record<string, DetailedStudent> {
    return Object.fromEntries(this.students.entries());
  }

  findManyByName(name: string): DetailedStudent[] {
    function normalize(str: string): string {
      return str.toLowerCase().replace(/[^a-z0-9]/g, "");
    }

    const finalName = normalize(name);

    return this.getStudentsWhere((student) => {
      return normalize(student.name).includes(finalName);
    });
  }

  static sortBySimilarity(value: string) {
    return (a: DetailedStudent, b: DetailedStudent) => {
      const aName = a.name.toLowerCase();
      const bName = b.name.toLowerCase();

      const finalValue = value.toLowerCase();

      const aSimilarity = similarity(aName, finalValue);
      const bSimilarity = similarity(bName, finalValue);

      return bSimilarity - aSimilarity;
    };
  }

  getByName(name: string): DetailedStudent | null {
    return (
      this.getStudents().find(
        (student) => student.name.toLowerCase() === name.toLowerCase(),
      ) ?? null
    );
  }

  getBaseVariants() {
    return this.getStudentsWhere((student) => student.baseVariantId === null);
  }

  getVariantsForBase(base: DetailedStudent) {
    const variants = [base];

    for (const student of this.getStudents()) {
      if (student.baseVariantId === base.id) {
        variants.push(student);
      }
    }

    return variants;
  }
}

export const studentContainer = new StudentContainer();

File: src/containers/students.rs
use crate::db::models::DetailedStudent;
use crate::utils::similarity::similarity;
use std::collections::HashMap;

#[derive(Debug, Default)]
pub struct StudentContainer {
    students: HashMap<String, DetailedStudent>,
}

impl StudentContainer {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn reload(&mut self, students: Vec<DetailedStudent>) {
        self.students.clear();
        for student in students {
            self.students.insert(student.id().to_string(), student);
        }
    }

    pub fn get_student(&self, key: &str) -> Option<&DetailedStudent> {
        self.students.get(key)
    }

    pub fn get_students(&self) -> Vec<&DetailedStudent> {
        self.students.values().collect()
    }

    pub fn get_students_where<F>(&self, predicate: F) -> Vec<&DetailedStudent>
    where
        F: Fn(&DetailedStudent) -> bool,
    {
        self.students.values().filter(|s| predicate(s)).collect()
    }

    pub fn find_many_by_name(&self, name: &str) -> Vec<&DetailedStudent> {
        fn normalize(s: &str) -> String {
            s.to_lowercase()
                .chars()
                .filter(|c| c.is_ascii_alphanumeric())
                .collect()
        }

        let final_name = normalize(name);
        self.get_students_where(|student| normalize(student.name()).contains(&final_name))
    }

    pub fn sort_by_similarity(value: &str, students: &mut Vec<&DetailedStudent>) {
        let final_value = value.to_lowercase();
        students.sort_by(|a, b| {
            let a_sim = similarity(&a.name().to_lowercase(), &final_value);
            let b_sim = similarity(&b.name().to_lowercase(), &final_value);
            b_sim
                .partial_cmp(&a_sim)
                .unwrap_or(std::cmp::Ordering::Equal)
        });
    }

    pub fn get_by_name(&self, name: &str) -> Option<&DetailedStudent> {
        self.students
            .values()
            .find(|s| s.name().eq_ignore_ascii_case(name))
    }

    pub fn get_base_variants(&self) -> Vec<&DetailedStudent> {
        self.get_students_where(|s| s.base_variant_id.is_none())
    }

    pub fn get_variants_for_base<'a>(&'a self, base: &'a DetailedStudent) -> Vec<&'a DetailedStudent> {
        let mut variants = vec![base];
        for student in self.students.values() {
            if student.base_variant_id.as_deref() == Some(base.id()) {
                variants.push(student);
            }
        }
        variants
    }
}
